cftime/_cftime.pyx

```diff
@@ -121,344 +121,163 @@ def _datesplit(timestr):
+                date = real_datetime(date.year, date.month, date.day, date.hour,
+                                     date.minute, date.second, date.microsecond)
+            # adjust for time zone offset
+            if getattr(date, 'tzinfo',None) is not None:
+                date = date.replace(tzinfo=None) - date.utcoffset()
+        else: # convert basedate and date to same calendar specific cftime.datetime instance
+            if not isinstance(basedate, DATE_TYPES[calendar]):
+                basedate =  to_calendar_specific_datetime(basedate, calendar, False)
+            if not isinstance(date, DATE_TYPES[calendar]):
+                date = to_calendar_specific_datetime(date, calendar, False)
+        if ismasked and not date:
+            times.append(None)
         else:
-            times = np.array(times, dtype='d')
-            shape = times.shape
-        dates = []
-        n = 0
-        for time in times.flat:
-            if ismasked and mask.flat[n]:
-                dates.append(None)
+            td = date - basedate
+            if factor == 1.0:
+                # units are microseconds, use integer division
+                # (fails on python 2.7)
+                times.append(td // timedelta(microseconds=1) )
             else:
-                # convert to total seconds
-                if unit in microsec_units:
-                    tsecs = time/1.e6
-                elif unit in millisec_units:
-                    tsecs = time/1.e3
-                elif unit in sec_units:
-                    tsecs = time
-                elif unit in min_units:
-                    tsecs = time*60.
-                elif unit in hr_units:
-                    tsecs = time*3600.
-                elif unit in day_units:
-                    tsecs = time*86400.
-                else:
-                    raise ValueError('unsupported time units')
-                # compute time delta.
-                days = tsecs // 86400.
-                msecsd = tsecs*1.e6 - days*86400.*1.e6
-                secs = msecsd // 1.e6
-                msecs = np.round(msecsd - secs*1.e6)
-                td = timedelta(days=days,seconds=secs,microseconds=msecs)
-                # add time delta to base date.
-                try:
-                    date = basedate + td
-                except OverflowError:
-                    msg="""
-OverflowError in python datetime, probably because year < datetime.MINYEAR"""
-                    raise ValueError(msg)
-                dates.append(date)
-            n += 1
-        if ismasked: # convert to masked array if input was masked array
-            dates = np.array(dates)
-            dates = np.ma.masked_where(dates==None,dates)
-            if isscalar:
-                return dates[0]
-            else:
-                return np.reshape(dates, shape)
+                #times.append( (td/timedelta(microseconds=1)) / factor )
+                # this works for python 2.7
```

Comment:
Could we switch to the Python 3 specific logic since we are now dropping Python 2.7 support?

Reply:
Done